### ISSUE TITLE
Suggestion regarding cached profile pictures

### ISSUE THREAD

**Issue by ahnaf-zamil** [Thenafi/meeseeks#1]
In the README it has been said
> No guarantee that the URL change will happen in time.

This is most likely due to the browser caching the profile picture as it is from the same URL if I am not wrong. But it's possible to make sure there is no cache stored in the browser and updated profile picture is pulled every single time.

If only there is no 3xx redirect system but rather the image request from the client is proxied through the backend (i.e the backend server sends the request on the user's behalf and returns the image to the user), the server may return a no-cache header such as `Cache-Control: no-cache` (Refer to [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) for this) and the browser will re-fetch for each request. As the backend is not caching or storing the pulled image into the application state, it may be possible to use this system for instant profile picture update with zero-cache.

One limitation I will say though is that it will put more load on the hosted server as the request is made on every single page load and for many users this will not scale well unless the backend is hosted in a high-performance server.

Does this mechanism seem plausible? Please let me know of your opinions and if there are any vulnerabilities or limitations in the above proposed idea.

**Comment by Thenafi** [comment]
Hey,

First of all, thank you for your suggestion. I appreciate it. I think my message in the readme was unclear, and I apologize for that. The lack of a guarantee is because the URL only changes when someone hits/requests it. So, if someone saves an option like "Every Hour," they might expect the URL to change at specific times, like 1:00 PM, 2:00 PM, and 3:00 PM. However, that's not how it works. The URL will only change when someone hits it. For instance, if someone visits the URL at 1:35 PM, it will remain the same until 2:35 PM, but after that, it will only change when someone hits it again. If nobody hits it, it will not change forever.

I hope that clarifies things. Also, there was a cache issue, but I resolved it by sending the [307 Temporary Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307), which works just fine and loads different images or URLs every time. Initially, I thought of storing and caching the images myself, but as you pointed out, it would be costly in terms of infrastructure and scalability. That's why I decided to keep it as simple and low-cost (free) as possible. I want the community to host and provide the facility to other users if many users like the idea 🤞.

When you pointed out the time guarantee issue, I gave it another thought. At first, I planned to use cron jobs or message queues, but then I abandoned the idea because it would add too much complexity to this simple project. However, as I write this, I think I can guarantee the time using simple math. 😂 It's funny that I didn't think of it earlier. I could easily calculate the hour and the URL for that time because I have both the time and the links data.
I hope you realize that I have found a solution for the lack of time guarantee issue.

Let me know your thoughts, and I will make time to solve it.

Again thank you so much.😊

**Comment by ahnaf-zamil** [comment]
Hello, thanks for reply!

So the guarantee issue was a misunderstanding on my side then. Glad you found a solution to this. I like the idea of using cron jobs or messsage queues but as you said, it would be complex. Although I do want to ask, what message/task queue were you planning to use? Redis, RabbitMQ, NATS? I'm curious.

Also, a way to implement the image change when it hits the specific hour would be to have a background thread that runs a check on the database every hour to see if any profile picture has "expired" its time or not. But Node.js is single threaded so I doubt it would be possible to do that. Yet, your solution is optimal for this case. I'm interested to see how far this project goes, it's truly a handy tool. 

Do you have any other plans for this or any features you plan on adding? I suggest adding a Docker compose file for those who may want to run a single command to start both the application container and a MongoDB instance together, what do you think?

**Comment by Thenafi** [comment]
Hey again,

I hope you're doing well. After doing some research and considering the size and scope of my project, I've decided that the Node.js queue package is sufficient for my needs. While Redis, RabbitMQ, and NATS provide advanced features, I don't think they're necessary for a small app like mine. These message brokers are designed for large-scale, distributed applications where message processing needs to be highly available, scalable, and fault-tolerant. Since my app is not even close to that scale, a simple local queue would suffice.

Regarding the image/URL changes, I have a plan on my sketch board to cache every URL in memory for a specific period, perhaps with background jobs. However, I don't want to put too much pressure on the free app I've already deployed, so I want to keep it as simple as possible. I do think that this simplicity would bring scalability a lot easier This way, others can deploy it on free servers if they want to. I think this app is just for fun, and I doubt anyone will use it besides TBD.

Lastly, I have to admit that I have little to no knowledge of Docker. While I'm not sure if I'll ever use it, I'd love to hear your contribution if you have experience with it. Thanks!
